File: nodejs/patterns/src/structural/facade.js
class Lights {
  dim() {
    return "lights dimmed";
  }
}

class Projector {
  on() {
    return "projector on";
  }
}

class SoundSystem {
  surround() {
    return "surround enabled";
  }
}

class HomeTheaterFacade {
  constructor(lights, projector, sound) {
    this.lights = lights;
    this.projector = projector;
    this.sound = sound;
  }

  watchMovie() {
    return [this.lights.dim(), this.projector.on(), this.sound.surround()].join(" | ");
  }
}

function demo() {
  return new HomeTheaterFacade(new Lights(), new Projector(), new SoundSystem()).watchMovie();
}

module.exports = { HomeTheaterFacade, demo };

File: python/patterns/src/prototype_patterns/structural/facade.py
"""Facade pattern example."""

from __future__ import annotations


class Lights:
    def dim(self) -> str:
        return "lights dimmed"


class Projector:
    def on(self) -> str:
        return "projector on"


class SoundSystem:
    def surround(self) -> str:
        return "surround enabled"


class HomeTheaterFacade:
    def __init__(self, lights: Lights, projector: Projector, sound: SoundSystem) -> None:
        self._lights = lights
        self._projector = projector
        self._sound = sound

    def watch_movie(self) -> str:
        actions = [self._lights.dim(), self._projector.on(), self._sound.surround()]
        return " | ".join(actions)


def demo() -> str:
    facade = HomeTheaterFacade(Lights(), Projector(), SoundSystem())
    return facade.watch_movie()
